_posts/2022-03-15-turn-based-offline-rl.md

```diff
@@ -0,0 +1,344 @@
+applied CQL with fixed alpha=0. For each run, CQL was fitted for 300 iterations.*
+
+## Conclusions and Future Work
+
+In this blog post we have presented a practical approach you could use to address cases where you have temporary
+and limited access to an environment, and you have computational resources at your disposal to train your RL algorithm
+“offline” only.
+
+In fact, the proposed solution falls in the middle between Online RL and Offline RL: our agent is warmed up
+by training it via Offline RL on a dataset generated by running the uniform random policy and then
+subsequently improved by accessing the environment in “turns”, thus partially simulating what you would get on a
+standard online RL scenario.
+
+In particular, we show that:
+
+* The turn-based procedure is effective since the policy learned in subsequent turns consistently improves as
+turns increase, matching the expectations. This result is demonstrated through some visualizations, showing how
+the agent chooses a better and faster path to target turn after turn.
+
+
+* The turn-based procedure allows getting an agent that is better than a random one, even after a small number of turns.
+The performance of the turn-based agent will be upper-bounded by the performance of the optimal policy.
+
+Moreover, we provide an easy-to-understand framework to prove the aforementioned hypothesis.
+
+Finally, we want to point out some limitations of our work that could be addressed as future work:
+
+* From one turn to the other, we pick the “best policy” as the one that achieves the best performance between all
+the runs via the “max” operator. This single policy is then propagated through the algorithm and used to generate
+the best new extension of the dataset. The inherent limitation of this approach is that by using the “max” we are
+not robust to the noise and we do not account for fluctuations in the performance of the Offline RL algorithm.
+A better approach would be aggregating policies by doing, for example, an ensemble of policies, and using this
+as the selected policy that is propagated forward in the algorithm.
+
+
+* Running a fixed Offline RL algorithm on a dataset that keeps changing its distribution of states and actions
+in time could be really challenging since a lot of algorithms in the literature require accurate hypertuning
+of the parameters. In future work, we would like to address this problem, proposing, for example, a way one
+could compute new hyper-parameters using the dataset size and some other properties as parameters for the computation.
+
+
+* One could argue that our hypothesis can work only on simplistic environments like GridWorlds.
+Even though we tested different configurations of grids, stressing more or less the algorithms,
+we admit that a more complete work would require the re-visitation of our hypothesis on a more diverse
+suite of environments. We plan to investigate this in the future.
```

Comment:
Bardzo ciekawy artykuł, a co więcej zrozumiały również dla osób niemających styczności z _RL_, czyli takimi jak ja 🙂 👍

Czy istnieje szansa, żebyście dodali link do repozytorium z kodem? Dzięki temu zainteresowany czytelnik mógłby dosłownie zagłębić się w temat i pobawić się różnymi parametrami bez odkrywania koła na nowo :wink:

Reply:
Thanks for the comment :-) I'm glad we succeeded in writing such an article! This was our goal!
About the link to the code, it's already linked in the text: [NeurIPS 2020 Offline RL Tutorial Colab Exercise](https://colab.research.google.com/drive/1oJOYlAIOl9d1JjlutPY66KmfPkwPCgEE?usp=sharing#scrollTo=4i64GqsO83mA). Is it enough for you or do you want me to make an explicit "References" section?